File: petclinic-java-gradle-react-tailwind-ts/src/main/java/dev/pollito/petclinic_java_gradle_react_tailwind_ts/vet/VetMapper.java
package dev.pollito.petclinic_java_gradle_react_tailwind_ts.vet;

import dev.pollito.petclinic_java_gradle_react_tailwind_ts.generated.model.VetDTO;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.specialty.Specialty;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.specialty.SpecialtyRepository;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.NotFoundException;
import java.util.HashSet;
import java.util.List;
import org.mapstruct.AfterMapping;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants;
import org.mapstruct.MappingTarget;
import org.mapstruct.ReportingPolicy;
import org.springframework.beans.factory.annotation.Autowired;

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING, unmappedTargetPolicy = ReportingPolicy.IGNORE)
public abstract class VetMapper {

    @Autowired
    protected SpecialtyRepository specialtyRepository;

    @Mapping(target = "vetSpecialtySpecialties", expression = "java(vet.getVetSpecialtySpecialties() == null ? java.util.List.of() : vet.getVetSpecialtySpecialties().stream().map(specialty -> specialty.getId()).toList())")
    public abstract VetDTO map(Vet vet);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "vetSpecialtySpecialties", ignore = true)
    public abstract Vet map(VetDTO vetDTO);

    @AfterMapping
    protected void afterMap(VetDTO vetDTO, @MappingTarget Vet vet) {
        final List<Specialty> vetSpecialtySpecialties = specialtyRepository.findAllById(
                vetDTO.getVetSpecialtySpecialties() == null ? List.of() : vetDTO.getVetSpecialtySpecialties());
        if (vetSpecialtySpecialties.size() != (vetDTO.getVetSpecialtySpecialties() == null ? 0
                : vetDTO.getVetSpecialtySpecialties().size())) {
            throw new NotFoundException("one of vetSpecialtySpecialties not found");
        }
        vet.setVetSpecialtySpecialties(new HashSet<>(vetSpecialtySpecialties));
    }
}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/vet/VetMapper.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.vet

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.specialty.SpecialtyRepository
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.NotFoundException
import org.mapstruct.AfterMapping
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingConstants
import org.mapstruct.MappingTarget
import org.mapstruct.ReportingPolicy
import org.springframework.beans.factory.annotation.Autowired

@Mapper(
    componentModel = MappingConstants.ComponentModel.SPRING,
    unmappedTargetPolicy = ReportingPolicy.IGNORE,
)
abstract class VetMapper {

    @Autowired
    protected lateinit var specialtyRepository: SpecialtyRepository

    @Mapping(
        target = "vetSpecialtySpecialties",
        expression = "java(vet.getVetSpecialtySpecialties() == null ? java.util.List.of() : vet.getVetSpecialtySpecialties().stream().map(specialty -> specialty.getId()).toList())",
    )
    abstract fun map(vet: Vet): VetDTO

    @Mapping(
        target = "id",
        ignore = true,
    )
    @Mapping(
        target = "vetSpecialtySpecialties",
        ignore = true,
    )
    abstract fun map(vetDTO: VetDTO): Vet

    @AfterMapping
    protected fun afterMap(vetDTO: VetDTO, @MappingTarget vet: Vet) {
        val vetSpecialtySpecialties = specialtyRepository.findAllById(
            vetDTO.vetSpecialtySpecialties
                ?: emptyList(),
        )
        if (vetSpecialtySpecialties.size != (
                if (vetDTO.vetSpecialtySpecialties == null) {
                    0
                } else {
                    vetDTO.vetSpecialtySpecialties!!.size
                }
                )
        ) {
            throw NotFoundException("one of vetSpecialtySpecialties not found")
        }
        vet.vetSpecialtySpecialties = vetSpecialtySpecialties.toMutableSet()
    }
}
